### PR TITLE
[SNOW-1518791] Change concat axis=1 from join_on_index to align_on_index behavior

### PR DESCRIPTION
**Pull request by sfc-gh-yzou**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

SNOW-1518791

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.
   - [ ] I acknowledge that I have ensured my changes to be thread-safe. Follow the link for more information: [Thread-safe Developer Guidelines](https://docs.google.com/document/d/162d_i4zZ2AfcGRXojj0jByt8EUq-DrSHPPnTa4QvwbA/edit#bookmark=id.e82u4nekq80k)

3. Please describe how your code solves the related issue.
concat when axis = 1 succeeded when the dataframe/series comes from the same dataframe with an align behavior. However, today we uses join which gives a different result when duplication occurs.
For example: with following dataframe
```
    C  A  D
2  1  a  3
1  2  b  2
2  3  c  1
```
pd.concat([df['C'], df['A']], axis = 1) gives 
```
    C  A
2  1  a
1  2  b
2  3  c
```
but snowpark pandas returns 
```
     C  A
2  1  a
2  1  c
1  2  b
2  3  a
2  3  c
```
In this pr, we fixed the behavior by switching to align behavior.
Also relaxed the align utility to enable customized sorting behavior for support the sort behavior of concat

### CODE REVIEW

File: src/snowflake/snowpark/modin/plugin/_typing.py [hunk 1]
```diff
@@ -54,6 +54,18 @@ class LabelIdentifierPair(NamedTuple):
     "coalesce",
 ]  # right and inner can also be supported if needed
 
+AlignSortLit = [
+    # Align operator provides a default sorting capability, which sort the
+    # align key lexicographically when the align type is outer, and the original
+    # dataframe is not aligned. No sort will happen for other align type.
```
Comment (sfc-gh-helmeleegy) [inline src/snowflake/snowpark/modin/plugin/_typing.py:60]:
```suggestion
    # dataframe is not aligned. No sort will happen for other align types.
```

File: src/snowflake/snowpark/modin/plugin/_typing.py [hunk 2]
```diff
@@ -54,6 +54,18 @@ class LabelIdentifierPair(NamedTuple):
...
+    # other types.
```
Comment (sfc-gh-helmeleegy) [inline src/snowflake/snowpark/modin/plugin/_typing.py:61]: remove?

**Review by sfc-gh-helmeleegy — APPROVED** [review]
LGTM, thanks!

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 3]
```diff
@@ -7351,13 +7351,17 @@ def concat(
...
+                # We rename index labels to make sure index columns are aligned level
                 # by level.
-                result_frame, _ = join_utils.join_on_index_columns(
-                    result_frame, other_frame, how=join, sort=sort
+                if sort is True:
+                    align_sort = "sort"
+                else:
+                    align_sort = "no_sort"
+                result_frame, _ = join_utils.align_on_index(
+                    result_frame, other_frame, how=join, sort=align_sort
```
Comment (sfc-gh-lmukhopadhyay) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:7364]: Is there no case for concat where sort="default_sort"?

File: CHANGELOG.md [hunk 4]
```diff
@@ -51,10 +51,12 @@
   - Treat an empty `subset` (e.g. `[]`) as if it specified all columns instead of no columns.
   - Raise a `TypeError` for a scalar `subset` instead of filtering on just that column.
   - Raise a `ValueError` for a `subset` of type `pandas.Index` instead of filtering on the columns in the index.
-
+- Fixed a bug when concat dataframe or series coming from the same dataframe when axis = 1.
```
Comment (sfc-gh-lmukhopadhyay) [inline CHANGELOG.md:54]:
```suggestion
- Fixed a bug when concat dataframe or series objects are coming from the same dataframe when axis = 1.
```

**Review by sfc-gh-nkumar — APPROVED** [review]
Thanks for fixing this.

File: CHANGELOG.md (hunk 4 above)
Comment (sfc-gh-yzou) [reply CHANGELOG.md]: updated

File: src/snowflake/snowpark/modin/plugin/_typing.py (hunk 1 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/modin/plugin/_typing.py]: updated

File: src/snowflake/snowpark/modin/plugin/_typing.py (hunk 2 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/modin/plugin/_typing.py]: Yes. Thanks for catching this.

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 3 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:7364]: for concat, there is no sort = "default_sort". However default_sort the default sorting approach for the utility, so all other usage of the align using the default_sort, like binary operation.

**Review by sfc-gh-lmukhopadhyay — APPROVED** [review]

**Pull request merged**